src/nvim/tui/tui.c

```diff
@@ -0,0 +1,847 @@
+               key.type == TERMKEY_TYPE_FUNCTION ||
+               key.type == TERMKEY_TYPE_KEYSYM) {
+      forward_modified_utf8(data->tk, &key);
+    } else if (key.type == TERMKEY_TYPE_MOUSE) {
+      forward_mouse_event(data->tk, &key);
+    }
+  }
+
+  if (result != TERMKEY_RES_AGAIN) {
+    return;
+  }
+
+  Integer ms = 0;
+  bool timeout = false;
+  // Check 'timeout' and 'ttimeout' to determine if we should send ESC
+  // after 'ttimeoutlen'. See :help 'ttimeout' for more information
+  Error err;
+  timeout = vim_get_option(cstr_as_string("timeout"), &err).data.boolean;
+  if (!timeout) {
+    timeout = vim_get_option(cstr_as_string("ttimeout"), &err).data.boolean;
+  }
+
+  if (timeout) {
+    ms = vim_get_option(cstr_as_string("ttimeoutlen"), &err).data.integer;
+  }
+
+  if (ms > 0) {
+    // Stop the current timer if already running
+    uv_timer_stop(&data->timer_handle);
+    uv_timer_start(&data->timer_handle, timer_cb, (uint32_t)ms, 0);
+  } else {
+    tk_getkeys(ui, true);
+  }
+}
+
+static void timer_cb(uv_timer_t *handle)
+{
+  tk_getkeys(handle->data, true);
+}
+
+static void poll_cb(uv_poll_t *handle, int status, int events)
+{
+  UI *ui = handle->data;
+  TUIData *data = ui->data;
+  TermKeyResult result;
+  if ((result = termkey_advisereadable(data->tk)) == TERMKEY_RES_AGAIN) {
+    tk_getkeys(ui, false);
+  } else if (result == TERMKEY_RES_ERROR) {
+    input_done();
+  }
+}
+
+static bool attrs_differ(HlAttrs a1, HlAttrs a2)
+{
+  return a1.foreground != a2.foreground || a1.background != a2.background
+    || a1.bold != a2.bold || a1.italic != a2.italic
+    || a1.undercurl != a2.undercurl || a1.underline != a2.underline
+    || a1.reverse != a2.reverse;
+}
+
+static void update_attrs(UI *ui, HlAttrs attrs)
+{
+  TUIData *data = ui->data;
+  unibi_out(ui, unibi_exit_attribute_mode, NULL);
+
+  data->params[0].i = attrs.foreground != -1 ? attrs.foreground : data->fg;
+  if (data->params[0].i != -1) {
+    unibi_out(ui, unibi_set_a_foreground, NULL);
+  }
+
+  data->params[0].i = attrs.background != -1 ? attrs.background : data->bg;
+  if (data->params[0].i != -1) {
+    unibi_out(ui, unibi_set_a_background, NULL);
+  }
+
+  if (attrs.bold) {
+    unibi_out(ui, unibi_enter_bold_mode, NULL);
+  }
+  if (attrs.italic) {
+    unibi_out(ui, unibi_enter_italics_mode, NULL);
+  }
+  if (attrs.underline) {
+    unibi_out(ui, unibi_enter_underline_mode, NULL);
+  }
+  if (attrs.reverse) {
+    unibi_out(ui, unibi_enter_reverse_mode, NULL);
+  }
+}
+
+static void print_cell(UI *ui, Cell *ptr)
+{
+  TUIData *data = ui->data;
+  if (attrs_differ(ptr->attrs, data->print_attrs)) {
+    update_attrs(ui, ptr->attrs);
+    data->print_attrs = ptr->attrs;
+  }
+  out(ui, ptr->data);
+}
+
+static void clear_region(UI *ui, int top, int bot, int left, int right,
+    bool repaint)
+{
+  TUIData *data = ui->data;
+  HlAttrs clear_attrs = EMPTY_ATTRS;
+  bool refresh_all = repaint && (right != ui->height - 1 || data->bg != -1);
```

Comment:
If the region to be cleared doesn't match the terminal right edge, eol_clr will delete more than it should